### PR TITLE
adds `Keyboard.dismiss()` function and document Keyboard

### PR DESCRIPTION
**Pull request by GantMan**
Simple and elegant.  Now someone can dismiss a keyboard in a way that makes sense.

```js
import { Keyboard } from 'react-native'

// Hide that keyboard!
Keyboard.dismiss()
```

+ docs

### CODE REVIEW

**Comment by facebook-github-bot-7** [comment]
By analyzing the blame information on this pull request, we identified @nicklockwood to be a potential reviewer.

**Comment by satya164** [comment]
Wow! I was planning to do this too. Thanks :D

**Comment by satya164** [comment]
cc @dmmiller

File: Libraries/Components/Keyboard/Keyboard.js [hunk 1]
```diff
@@ -13,5 +13,10 @@
 
 const NativeEventEmitter = require('NativeEventEmitter');
 const KeyboardObserver = require('NativeModules').KeyboardObserver;
+var dismissKeyboard = require('dismissKeyboard');
```
Comment (satya164) [inline Libraries/Components/Keyboard/Keyboard.js]: nit: change it to `const`

File: Libraries/Components/Keyboard/Keyboard.js [hunk 2]
```diff
@@ -13,5 +13,10 @@
...
 
-module.exports = new NativeEventEmitter(KeyboardObserver);
+let keyboard = new NativeEventEmitter(KeyboardObserver);
```
Comment (satya164) [inline Libraries/Components/Keyboard/Keyboard.js]: nit: change to `const`

**Comment by facebook-github-bot-9** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/48d98688534c434e6828e95e94d7dced06c4a4bc..8cbf3fa83c934e1a89e423d0c8e219cdd5fbd145)

File: Libraries/Components/Keyboard/Keyboard.js (hunk 2 above)
Comment (GantMan) [inline Libraries/Components/Keyboard/Keyboard.js]: Would that be correct?  I'm adding a function to `keyboard`.  Would that be defeating the idea of `const`?

Comment (satya164) [inline Libraries/Components/Keyboard/Keyboard.js]: `const` just creates an immutable binding to the variable. The value is still mutable.

Comment (satya164) [inline Libraries/Components/Keyboard/Keyboard.js]: also nit: capitalize `Keyboard`

Comment (GantMan) [inline Libraries/Components/Keyboard/Keyboard.js]: You're the boss! :)

**Comment by facebook-github-bot-8** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/8cbf3fa83c934e1a89e423d0c8e219cdd5fbd145..58813d50d78fc5e77ca8a09c25c94cac2e6d876c)

**Comment by facebook-github-bot-9** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/58813d50d78fc5e77ca8a09c25c94cac2e6d876c..55254f26253d8aa931dae8f1665a80915303e02e)

**Comment by satya164** [comment]
Adding docs will be awesome too!

**Comment by GantMan** [comment]
@satya164  I agree!  

Where would the docs for that be?  Would I start documenting the `Keyboard` component, and that would in turn generate the page?

**Comment by satya164** [comment]
@GantMan We'll need to add comments in `Keyboard.js` and add the file to `extractDocs.js`

**Comment by GantMan** [comment]
on it.

**Comment by facebook-github-bot-3** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/55254f26253d8aa931dae8f1665a80915303e02e..7a81d493a3864e4edffabb904acbc34ce2fcd27f)

**Comment by facebook-github-bot-6** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/7a81d493a3864e4edffabb904acbc34ce2fcd27f..ab3bbd34a43ba622589441f520de37311961a2ce)

**Comment by GantMan** [comment]
@satya164  - I have done the first draft of documenting the Keyboard component.  Unfortunately, I was unable to use some of the tricks you guys have in other docs because the class isn't exposed, but I believe what i have does a fair job at documenting regarding the structure in this file.

Please let me know what updates I need :)

**Comment by facebook-github-bot-5** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/ab3bbd34a43ba622589441f520de37311961a2ce..7a81d493a3864e4edffabb904acbc34ce2fcd27f)

**Comment by satya164** [comment]
Awesome. Please post a screenshot of the docs :D

**Comment by GantMan** [comment]
Took me a second to figure out the docs, I'm not familiar with `react-docgen`.  The docgen would only read the file if it was a component.   So to make this work, I had to mock a component to have documentation.  I think all the headache stems from the concept that Keyboard is a component when it probably should have been a Util.

Here's a the gif of the generated page:
keyboard

**Comment by facebook-github-bot-1** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/bd0b44b0b308fd481f5f30e2c1f500c0da8b9687..b908fc39b20230c2302257e77a4561fa6b889e2d)

**Comment by satya164** [comment]
@GantMan awesome. BTW circle CI is failing. Has to do something with extracting docs. Probably we need to write the full class.

File: Libraries/Components/Keyboard/Keyboard.js [hunk 3]
```diff
@@ -11,7 +11,94 @@
...
+ *     return (
+ *       <TextInput
+ *         onSubmitEditing={Keyboard.dismiss}
+ *       />
+ *     );
+ *   }
+ * }
+ *```
+ */
+class DocsKeyboard extends React.Component {
```
Comment (satya164) [inline Libraries/Components/Keyboard/Keyboard.js]: You don't need to extend `React.Component`

File: Libraries/Components/Keyboard/Keyboard.js [hunk 4]
```diff
@@ -11,7 +11,94 @@
...
+   *can be any of the following
+   * - `keyboardWillShow`
+   * - `keyboardDidShow`
+   * - `keyboardWillHide`
+   * - `keyboardDidHide`
+   * - `keyboardWillChangeFrame`
+   * - `keyboardDidChangeFrame`
+   * @param {function} jsFunction function to be called when the event fires.
+   */
+  addListener (nativeEvent, jsFunction) {
```
Comment (satya164) [inline Libraries/Components/Keyboard/Keyboard.js]: You also need to add `removeAllListeners` and `removeSubscription` - https://github.com/facebook/react-native/blob/master/Libraries/EventEmitter/NativeEventEmitter.js

**Comment by GantMan** [comment]
Looks like Circle CI ended up working  ¯\_(ツ)_/¯

Here's the updated docs:
image

**Comment by facebook-github-bot-3** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/b908fc39b20230c2302257e77a4561fa6b889e2d..8cf97bce915ddfc0499e396f3190a820c68618cd)

**Comment by facebook-github-bot-4** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/8cf97bce915ddfc0499e396f3190a820c68618cd..df4da15754fd4756c47e8b9a8a8a26bdc863ed82)

**Comment by GantMan** [comment]
Also @satya164 - thanks for your patience as I learn how to add to the docs.  That's something I can really start to help with going forward 👍

**Comment by facebook-github-bot-2** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/df4da15754fd4756c47e8b9a8a8a26bdc863ed82..def72f612fa483c50c4bb3ffc79cc86ea0eceb05)

**Comment by satya164** [comment]
No problem :)

BTW tests are failing. Can you rebase?

**Comment by facebook-github-bot-6** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/def72f612fa483c50c4bb3ffc79cc86ea0eceb05..f6e4511eeab7a30e7eb1fa6719d5e406aede76e2)

**Comment by GantMan** [comment]
@satya164 - Is this circle CI error related to my code?  I'm not getting how a CI is failing.

**Comment by GantMan** [comment]
@satya164  - tests are good.  Need anything else for this one?

**Comment by satya164** [comment]
Can you remove the propTypes and render function?

**Comment by satya164** [comment]
Also I'd change the naming to use 'module' instead of component

**Comment by GantMan** [comment]
@satya164  - I believe I tried removing the render, and then the docs errored.  I can try again, to be 100% sure.  I would have not wanted it, either.  But without render I think the react-docgen fails.  Checking now.

**Comment by GantMan** [comment]
@satya164  - Both of those seem to be required by docgen... unless you know the magic to remove them.

removing proptypes causes `Error: TypeError: Cannot convert undefined or null to object`
removing render causes `500 Error: No suitable component definition found.`

**Comment by GantMan** [comment]
I don't want to start editing the docgen resolver, and that seems to be the fix that people say (from Googling).  Open to feedback on how to proceed.  Would be a shame to scrap these docs, but if that's what happens, I'm ok with it.

**Comment by satya164** [comment]
cc @hramos any idea on how the docs are generated?

**Comment by facebook-github-bot-6** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/157f4e4d00a7ed9d87d7af41ddf5188f07af47d1..4d2ec5c020dc963c255eca22ef032f0c98870f2a)

**Comment by hramos** [comment]
This is what I typically do when I run into docgen issues:

1. Try running react-docgen against the one file, to isolate the error being thrown by the one file from any other docgen issues.
2. If react-docgen fails on the file, verify if there's a newer version of react-docgen I can try (compared to the one we're using in the website/package.json).
3. If that still fails, file a bug report against react-docgen as needed.

**Comment by GantMan** [comment]
@hramos - can you give us your opinion real quick?   I wouldn't call this an error, it's a bit of a corner case.

We want docs on something that is coming from native module.  Therefore it doesn't fit the normal React class structure.  The resolvers work great, it's just there should be no propTypes or render method.

**Comment by hramos** [comment]
It sounds like the docgen works if you include the render/propTypes methods, would that block you from merging? Or are you proposing that react-docgen be updated to not require these?

**Comment by satya164** [comment]
@hramos It's not a React component, so `render` and `propTypes` shouldn't be there.

Will changing it to some other format work? e.g. an object instead of a class?

**Comment by hramos** [comment]
That I don't know - I'd refer to the other APIs currently rendered by react-docgen in the RN website and see if this is an issue that has been worked around before.

**Comment by satya164** [comment]
@GantMan this is a plain class and seems to be working with docs? https://github.com/facebook/react-native/blob/master/Libraries/AppState/AppState.js

**Comment by GantMan** [comment]
Ok @satya164 -  Here's the info!  

### The TL:DR;
AppState is an API, and therefore brought in at a different level for `extractDocs.js` with different rules.
Keyboard is a component, and held to component doc rules....

#### Now, that being said.
I don't think Keyboard should have been a component.  As we both see, it really is more of an API.  It behaves like AppState.js but here's the problem with adding it to the API section for documentation purposes.
1. Technically it's not (yet)
2. The API is documented by what's exported... so our documentation class won't work if it's not what gets exported as default.  We must export what is actively used for the docs to work.

#### Options going forward
1. We can document it as a component for now, since that's technically what it is.  That means the existing PR is good to go.   We can then open a ticket requesting to move it to an API.  Which I would probably be happy to do... but it might have it's own battles.  Separate concern in my opinion.
2. We could move this to an API under this PR.. but it's getting pretty lengthy/noisy
3. We could abandon the docs (not a happy option)

Let me know how we should proceed.

**Comment by hramos** [comment]
@GantMan thanks for the detailed writeup. I'm happy with #1.

**Comment by satya164** [comment]
> Technically it's not (yet)

Why do you think it's not an API yet?

> The API is documented by what's exported... so our documentation class won't work if it's not what gets exported as default.

Let's just export the class.

> We can document it as a component for now, since that's technically what it is. 

It's not a React component. Rendering it will do nothing. Technically it's not a component either. It's just a class with the render method and a static `propTypes` property.

I prefer to go with 2. Since it's wrong to document something as component where clearly it's not.

Here is what we need to do,

1. Remove the line `Keyboard.dismiss = dismissKeyboard`
2. Change `class DocsKeyboard {` to `module.exports = class Keyboard`
3. Remove `propTypes` and `render`
4. Remove the line `module.exports = Keyboard`

The change it pretty simple and straightforward. So let's do it and document this properly.

**Comment by GantMan** [comment]
@satya164  -   Everything you said sounds good, I just have **1 quick question** to discuss.

#### 1. API vs Component

> Why do you think it's not an API yet?

Well, it's in the components folder:  `Libraries/Components/Keyboard/Keyboard.js`.  If it were an API shouldn't it be in the Libraries folder (like `AppState.js` is)?  Do you want me to move it to the Libraries folder, or are we planning on keeping it in Components?   (_n.b. Moving to Libraries would mean changing all import/require references._)

Once we're square on the above, I can make the requested modifications for this PR 😄 👍

**Comment by facebook-github-bot-1** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/4d2ec5c020dc963c255eca22ef032f0c98870f2a..05f12b09883f0e20b239c8e100e8422089c2eb0c)

**Comment by facebook-github-bot-9** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/05f12b09883f0e20b239c8e100e8422089c2eb0c..91535e2ff78db7ccfe89812c4a76cfbaa8500a82)

**Comment by GantMan** [comment]
@satya164  - DONE!

**Bonus:** Added flow types

One of the downsides to moving to API, is that the `@params` don't seem to work in the documentor.   You can see this already existing in things like [Clipboard](http://facebook.github.io/react-native/releases/0.34/docs/clipboard.html#clipboard) on the live site.  I think it still imparts important information, though.  Definitely a different ticket.

File: Libraries/Components/Keyboard/Keyboard.js [hunk 5]
```diff
@@ -13,5 +13,103 @@
...
+ *     return (
+ *       <TextInput
+ *         onSubmitEditing={Keyboard.dismiss}
+ *       />
+ *     );
+ *   }
+ * }
+ *```
+ */
+class Keyboard {
```
Comment (satya164) [inline Libraries/Components/Keyboard/Keyboard.js]: Let's just `export default class Keyboard`

File: Libraries/Components/Keyboard/Keyboard.js [hunk 6]
```diff
@@ -13,5 +13,103 @@
...
+   * - `keyboardWillShow`
+   * - `keyboardDidShow`
+   * - `keyboardWillHide`
+   * - `keyboardDidHide`
+   * - `keyboardWillChangeFrame`
+   * - `keyboardDidChangeFrame`
+   *
+   * @param {function} jsFunction function to be called when the event fires.
+   */
+  addListener (nativeEvent: string, jsFunction: Function) {
```
Comment (satya164) [inline Libraries/Components/Keyboard/Keyboard.js]: Change to `static addListener`

**Comment by satya164** [comment]
Sorry for dragging it further. But could you export the class itself instead of exporting an instance and make the methods static?

Thanks a lot for not abandoning it. Much appreciated

**Comment by GantMan** [comment]
I don't mind @satya164 -  Each iteration gets better!

I agree with you, static methods would be better, but `react-docgen` doesn't seem to work with static methods.  When I used static methods the resulting documentation page became a 500 - complaining there were no component methods for it to use:  
image

If you or @hramos know how to make static methods work, I'd be happy to switch it!

**Comment by satya164** [comment]
@GantMan Oh. Okay then. Let's just use a plain object instead of a class like [Clipboard.](https://github.com/facebook/react-native/blob/master/Libraries/Components/Clipboard/Clipboard.js) does then?

**Comment by facebook-github-bot-9** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/91535e2ff78db7ccfe89812c4a76cfbaa8500a82..ccca8fad375fc32c59f7b470a0d042aefe1d580c)

**Comment by facebook-github-bot-8** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/91535e2ff78db7ccfe89812c4a76cfbaa8500a82..ccca8fad375fc32c59f7b470a0d042aefe1d580c)

**Comment by facebook-github-bot-9** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/ccca8fad375fc32c59f7b470a0d042aefe1d580c..332ea361bd01e40cbd3168241122a564393d6c02)

**Comment by GantMan** [comment]
@satya164  - Done!  👍  Ready to roll

**Comment by facebook-github-bot-2** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/332ea361bd01e40cbd3168241122a564393d6c02..01201af8eb4301a53a3bf60eb40be338f5719fdf)

**Comment by satya164** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-5** [comment]
Thanks for importing. If you are a Facebook employee, you can view this diff [on Phabricator](https://phabricator.intern.facebook.com/D3935357).

**Comment by facebook-github-bot-6** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

**Comment by facebook-github-bot-3** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9925/files/01201af8eb4301a53a3bf60eb40be338f5719fdf..0e17bdb855e1e9c440e747cd949853a3438f7afb) - [changes since last import](https://github.com/facebook/react-native/pull/9925/files/01201af8eb4301a53a3bf60eb40be338f5719fdf..0e17bdb855e1e9c440e747cd949853a3438f7afb)

**Comment by GantMan** [comment]
Been seeing a lot of failed imports lately.  I hope it's not just me.

**Comment by GantMan** [comment]
all CI looks good @satya164 - any idea why import failed?

**Comment by satya164** [comment]
@facebook-github-bot shipit

**Comment by sumzer** [comment]
Hi,
I've problem on React-Native v0.35 with this method.
I've got error:
`_reactNative.Keyboard.dismiss is not a function`

Best regards

**Comment by GantMan** [comment]
Hey @sumzer - as you can see, it's slated for 0.36.0 in release candidate
https://github.com/facebook/react-native/releases/tag/v0.36.0-rc.0

**Comment by derekgreenberg** [comment]
@sumzer - in the meantime, add this to the top of your file:

`import dismissKeyboard from 'dismissKeyboard'`

Then you can reference it like this:

`dismissKeyboard()`

This works in RN 0.34.1

**Comment by pstanton** [comment]
This implementation requires the focus still be in a TextInput component. I had the scenario where I'd done a refresh while focussed - the app reloaded without a TextInput, but the keyboard remained. In this case there was no way to remove the keyboard, ie calling `Keyboard.dismiss()` had zero impact.

from https://github.com/facebook/react-native/blob/v0.56.0-rc.2/Libraries/Utilities/dismissKeyboard.js :
```
function dismissKeyboard() {
  TextInputState.blurTextInput(TextInputState.currentlyFocusedField());
}
```
